guidance/library/_json.py

```diff
@@ -566,27 +566,45 @@ def object(
         properties: Mapping[str, JSONSchema],
         additional_properties: JSONSchema,
         required: Sequence[str],
+        base_uri: str,
     ):
-        # "required" keys will be validated against "properties" if they're present, otherwise against "additionalProperties".
-        # If "additionalProperties" is False, then required keys must be in "properties".
-        if any(k not in properties for k in required) and additional_properties is False:
-            raise ValueError(
-                f"Required properties not in properties but additionalProperties is False."
-                f" Missing required properties: {list(r for r in required if r not in properties)}"
-            )
+        illegal_keys = set()
+        property_grammars: dict[str, GrammarFunction] = {}
+        for name, schema in properties.items():
+            try:
+                property_grammars[name] = self.json(json_schema=schema, base_uri=base_uri)
+            except UnsatisfiableSchemaError as e:
+                # We get here if the schema is a literal False or is otherwise determined to be unsatisfiable
+                if name in required:
+                    raise UnsatisfiableSchemaError(f"Required property {name!r} is unsatisfiable") from e
+                # Use json_dumps to properly quote / escape the key
+                key = json_dumps(name)
+                illegal_keys.add(key)
+
+        additional_properties_grammar: Optional[GrammarFunction] = None
+        try:
+            additional_properties_grammar = self.json(json_schema=additional_properties, base_uri=base_uri)
+        except UnsatisfiableSchemaError as e:
+            if any(k not in properties for k in required):
```

Comment:
No `else` needed here? even to just re-raise (and in other exception blocks with a conditional)

Reply:
OK. If this and other places this sort of pattern pops up are considered normal. Perhaps a comment, though?

Reply:
Happy to add a comment